examples/diagram.dx

```diff
@@ -90,63 +94,79 @@ def updateGeom (update: GeomStyle -> GeomStyle) (d:Diagram) : Diagram =
 def setFillColor   (c:HtmlColor) : Diagram -> Diagram = updateGeom $ setAt #fillColor   (Just c)
 def setStrokeColor (c:HtmlColor) : Diagram -> Diagram = updateGeom $ setAt #strokeColor (Just c)
 def setStrokeWidth (w:Int)       : Diagram -> Diagram = updateGeom $ setAt #strokeWidth w
--- TODO: argumentless def!
-removeStroke : Diagram -> Diagram = updateGeom $ setAt #strokeColor Nothing
-removeFill   : Diagram -> Diagram = updateGeom $ setAt #fillColor   Nothing
+def removeStroke                 : Diagram -> Diagram = updateGeom $ setAt #strokeColor Nothing
+def removeFill                   : Diagram -> Diagram = updateGeom $ setAt #fillColor   Nothing
 
 '## Serialization to SVG string
 
-def quote (s:String) : String = str "\"" <> s <> str "\""
+'Non-inlinable versions to improve compile times. (Non-inlined functions have to be monomorphic right now).
+
+@noinline
+def strCatUncurried ((xs,ys):(String & String)) : String = xs <> ys
+
+def (<.>) (xs:String) (ys:String) : String = strCatUncurried (xs, ys)
```

Comment:
Is the uncurried version needed because we can't partially apply a noinline function? If so, a few questions about that:

- What happens if you forget this, do you get an understandable error?
- Could we do this transformation automatically (either in the parser or as a lowering pass) so that users don't have to think about this? Maybe that's for a future PR, though.

Reply:
Right, it has to be uncurried because noinline only works on truly first-order functions. As you point out, curried functions are only higher-order in a narrow pedantic sense, and there's a simple curry/uncurry transformation. We should definitely do that transformation automatically. The errors are likely very bad if you break the contract, and we should improve those too. I'm thinking of the decorator as mostly a stop-gap measure before we add some sort of cost model to the compiler so that it can decide itself when to inline.